src/main/kotlin/dev/minn/jda/ktx/JDABuilder.kt

```diff
@@ -0,0 +1,399 @@
+        }
+
+    var enableShutdownHook: Boolean = true
+        set(value) {
+            builder.setEnableShutdownHook(value)
+            field = value
+        }
+
+    var autoReconnect: Boolean = true
+        set(value) {
+            builder.setAutoReconnect(value)
+            field = value
+        }
+
+    var eventManager: IEventManager? = null
+        set(value) {
+            builder.setEventManager(value)
+            field = value
+        }
+
+    var audioSendFactory: IAudioSendFactory? = null
+        set(value) {
+            builder.setAudioSendFactory(value)
+            field = value
+        }
+
+    var idle: Boolean = false
+        set(value) {
+            builder.setIdle(value)
+        }
+
+    var activity: Activity? = null
+        set(value) {
+            builder.setActivity(value)
+            field = value
+        }
+
+    var status: OnlineStatus? = null
+        set(value) {
+            if (value != null) {
+                builder.setStatus(value)
+            }
+            field = value
+        }
+
+    var eventListeners: MutableSet<Any> = mutableSetOf()
+
+    var maxReconnectDelay: Int = 900
+        set(value) {
+            builder.setMaxReconnectDelay(value)
+            field = value
+        }
+
+    inline fun sharding(build: InlineSharding.() -> Unit) {
+        val sharding = InlineSharding().also(build)
+        builder.useSharding(sharding.shardId, sharding.shardTotal)
+    }
+
+    var sessionController: SessionController? = null
+        set(value) {
+            builder.setSessionController(value)
+            field = value
+        }
+
+    var voiceDispatchInterceptor: VoiceDispatchInterceptor? = null
+        set(value) {
+            builder.setVoiceDispatchInterceptor(value)
+            field = value
+        }
+
+    var chunkingFilter: ChunkingFilter? = null
+        set(value) {
+            builder.setChunkingFilter(value)
+            field = value
+        }
+
+    var gatewayIntents: MutableSet<GatewayIntent> = mutableSetOf()
+
+    inline fun enableIntents(cacheFlag: GatewayIntent, vararg cacheFlags: GatewayIntent) {
+        builder.enableIntents(cacheFlag, *cacheFlags)
+    }
+
+    inline fun enableIntents(cacheFlags: Set<GatewayIntent>) {
+        builder.enableIntents(cacheFlags)
+    }
+
+    inline fun disableIntents(cacheFlag: GatewayIntent, vararg cacheFlags: GatewayIntent) {
+        builder.disableIntents(cacheFlag, *cacheFlags)
+    }
+
+    inline fun disableIntents(cacheFlags: Set<GatewayIntent>) {
+        builder.disableIntents(cacheFlags)
+    }
+
+    var largeThreshold: Int = 250
+        set(value) {
+            builder.setLargeThreshold(value)
+            field = value
+        }
+
+    var maxBufferSize: Int = 2048
+        set(value) {
+            builder.setMaxBufferSize(value)
+            field = value
+        }
+
+    var injectKtx: Boolean = true
+
+    class InlineSharding {
+        var shardId: Int = 0
+        var shardTotal: Int = 0
```

Comment:
This default makes no sense

Reply:
By default, JDA only uses 1 shard with an id of 0 (Actually, should that `shardTotal` be `1`? I might have overlooked that)
but if anyone _is_ using that, then they basically _have_ to assign both of those values. The default is just to get around the compiler warnings. Though, `lateinit` could be used, but I'm not sure how you feel about that.

Reply:
Looking at some more stuff, it might be better to just provide a builder for `DefaultShardManager`.
Because if you're using a single shard, it functions the same way as a normal JDA instance.
Although, there might be some reasons that one might want to construct a `JDA` instance instead of a `DefaultShardManager`.

I'm open to suggestions.

Reply:
I think it would be better to just use a method with default values like i suggested in the other comment. The pattern here doesn't really make sense here since you just set an id essentially.

Reply:
I saw that.

Actually, thinking about it, it might be better to just have a single method, without any default parameters, that takes in both of the values for it.

Reply:
Actually actually, it might even be a bit nicer if we instead use a variable of type `Pair<Int, Int>` and it'd be used like this:

```kotlin
val builder = DefaultJDABuilder("token") {
    // stuff
}

for (id in 0..10) {
    builder.sharding = id to 10
    builder.build()
}
```